hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -855,12 +1080,128 @@ public void breakLease(final Path path, final TracingContext tracingContext) thr
     client.breakLease(getRelativePath(path), tracingContext);
   }
 
-  public void rename(final Path source, final Path destination, TracingContext tracingContext) throws
-          AzureBlobFileSystemException {
+  public void rename(final Path source, final Path destination,
+      final AzureBlobFileSystem azureBlobFileSystem, TracingContext tracingContext) throws
+          IOException {
     final Instant startAggregate = abfsPerfTracker.getLatencyInstant();
     long countAggregate = 0;
     boolean shouldContinue;
 
+    if (getAbfsConfiguration().getPrefixMode() == PrefixMode.BLOB) {
+      LOG.debug("Rename for src: {} dst: {} for non-HNS blob-endpoint",
+          source, destination);
+      final Boolean isSrcExist;
+      final Boolean isSrcDir;
+      /*
+      * Fetch the list of blobs in the given sourcePath.
+      */
+      List<BlobProperty> srcBlobProperties = getListBlobs(source,
+          tracingContext, null, null, true);
+      final BlobProperty blobPropOnSrc;
+      if (srcBlobProperties.size() > 0) {
+        LOG.debug("src {} exists and is a directory", source);
+        isSrcExist = true;
+        isSrcDir = true;
+        /*
+        * Fetch if there is a marker-blob for the source blob.
+        */
+        BlobProperty blobPropOnSrcNullable = getBlobPropertyWithNotFoundHandling(source, tracingContext);
+        if(blobPropOnSrcNullable == null) {
+          /*
+          * There is no marker-blob, the client has to create marker blob before
+          * starting the rename.
+          */
+          //create marker file; add in srcBlobProperties;
+          LOG.debug("Source {} is a directory but there is no marker-blob", source);
+          azureBlobFileSystem.create(source);
```

Comment:
It is not right to pass the entire filesystem instance inside this method. This check should be done in the filesystem rename method only.

Reply:
Yes, filesystem in store is not good. Have refactored it.

Reply:
Here we will call the createDirectory() method of store.